src/components/SingleElement.tsx

```diff
@@ -0,0 +1,83 @@
+import { useEffect, useState } from 'react';
+import { Element } from '../types/Common';
+import { transformCoordinates } from '../assets/utilities';
+
+interface Props {
+  element: Element;
+  setElements: React.Dispatch<React.SetStateAction<Element[]>>;
+  svgContainerRef: SVGSVGElement | null;
+}
+
+const SingleElement = ({ element, setElements, svgContainerRef }: Props) => {
+  const [isDragging, setIsDragging] = useState<boolean>(false);
+  const [startMovePosition, setStartMovePosition] = useState<{
+    x: number;
+    y: number;
+  }>({ x: 0, y: 0 });
+
+  const handleMouseDown = (
+    // maybe need complex type
+    e:
+      | React.MouseEvent<SVGRectElement, MouseEvent>
+      | React.MouseEvent<SVGCircleElement, MouseEvent>
+      | React.MouseEvent<SVGLineElement, MouseEvent>,
+  ) => {
+    const { transX, transY } = transformCoordinates(
+      svgContainerRef,
+      e.clientX,
+      e.clientY,
+    );
+    setStartMovePosition({ x: transX, y: transY });
+    setIsDragging(true);
+  };
+
+  const handleMouseMove = (e: MouseEvent) => {
+    if (isDragging && element.x !== undefined && element.y !== undefined) {
+      const { transX, transY } = transformCoordinates(
+        svgContainerRef,
+        e.clientX,
+        e.clientY,
+      );
+      const newX = transX - startMovePosition.x + element.x;
+      const newY = transY - startMovePosition.y + element.y;
+      setElements((prev) =>
+        prev.map((el) =>
+          el.id === element.id ? { ...el, x: newX, y: newY } : el,
+        ),
+      );
+    }
+  };
+
+  const handleMouseUp = () => {
+    setIsDragging(false);
+  };
+
+  // Add event listeners for mouse move and mouse up when dragging
+  useEffect(() => {
+    if (isDragging) {
+      window.addEventListener('mousemove', handleMouseMove);
+      window.addEventListener('mouseup', handleMouseUp);
+    } else {
+      window.removeEventListener('mousemove', handleMouseMove);
+      window.removeEventListener('mouseup', handleMouseUp);
+    }
+
+    return () => {
+      window.removeEventListener('mousemove', handleMouseMove);
+      window.removeEventListener('mouseup', handleMouseUp);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
why are you disabling the deps check?

Reply:
because of this
![image](https://github.com/KirillSerg/drawler/assets/99827460/32a93fc1-daaf-47ca-9ac2-dc81da0cdf0c)

or I need like it says?

Reply:
when i do like it says it broke my app
I think i need to change this fn on every re-render

Reply:
I don't think this will break anything if 
```js
const handleMouseUp = () => {
    setIsDragging(false);
  };
```

will be moved into the hook itself

it's generally a bad practice to disable deps check, since it will introduce an error, instead you should figure out why it is a problem and how to fix it

Reply:
problem in moving handleMouseDown in useEffect or making a useCallback

Reply:
problem in moving handleMouseMove in useEffect or making a useCallback

Reply:
`useCallback` is not necessary here, moving the function into the effect will work just fine

Reply:
when I move the ```handleMouseMove``` to ```useEffect```, do I need to add to the ```useEffect``` dependency everything that will be in this ```handleMouseMove```?
Besides, when I implement Jotai, the use effect will become unnecessary and this question will disappear by itself

Reply:
++